aten/src/ATen/native/TensorIterator.cpp

```diff
@@ -73,177 +73,278 @@ void TensorIterator::reorder_dimensions() {
+      }
     }
+    return;
   }
 
-  for (auto &op : operands_) {
-    bool skip_output = compute_common_dtype_only_for_inputs && op.is_output;
-    bool is_different = op.tensor.defined() && op.current_dtype != common_dtype_;
+  // CPU
+  //   - Updates output data if any outputs are undefined (see allocate_outputs())
+  //   - Checks that all tensors are on the same device, if requested
+  //   - Checks that the output can be safely cast to, if requested
+  //   - Creates temporaries for inputs that are not the common dtype if
+  //       promote_inputs_to_common_dtype_ is true
+  //   - Creates temporaries for outputs that are not the common dtype if
+  //       cast_common_dtype_to_outputs_ is true
+  // NOTE: for inplace operations the temporary created for the input is
+  //   reused for the output, when available
+  for (auto& op : operands_) {
+    if (!op.is_type_defined()) {
+      op.target_dtype = common_dtype_;
+      continue;
+    }
 
-    if (may_have_differing_types) {
-      validate_dtype(op, common_dtype_, common_dtype_strategy_);
-      bool cast_by_copy = compute_common_dtype && !common_device_is_cuda && !skip_output;
-      if (cast_by_copy) {
-        maybe_copy_casting_to_common_dtype(op, common_dtype_);
-      }
+    // Skips undefined tensors
+    if (!op.tensor.defined()) {
+      continue;
     }
 
-    // Checks for tensors on the wrong device
-    if (op.tensor.defined() && op.device != op.tensor.device()) {
+    // Checks same device
+    if (check_all_same_device_) {
       if (op.is_output) {
-        TORCH_CHECK(false, "output with device ", op.tensor.device(),
-                    " doesn't match the desired device ", op.device);
+        TORCH_CHECK(op.device == common_device, "output with device ", op.tensor.device(),
+                    " doesn't match the desired device ", common_device);
       } else {
-        TORCH_CHECK(false, "expected device ", op.device,
+        TORCH_CHECK(op.device == common_device, "expected device ", common_device,
                     " but got device ", op.tensor.device());
       }
     }
+
+    // Checks safe casting
+    if (op.is_output && op.current_dtype != common_dtype_ && enforce_safe_casting_to_output_) {
+        TORCH_CHECK(canCast(common_dtype_, op.current_dtype),
+                    "result type ", common_dtype_, " can't be cast to the "
+                    "desired output type ", op.current_dtype);
+    }
+
+    // Casts to outputs by creating temporaries of the correct dtype (if needed)
+    // NOTE: for inplace operations this reuses the copy made when promoting
+    //   inputs, when available
+    if (op.is_output && op.current_dtype != common_dtype_ && cast_common_dtype_to_outputs_ &&
+        !(op.is_read_write && promote_inputs_to_common_dtype_)) {
+      op.tensor = at::empty_like(op.tensor,
+                                 op.tensor.options().dtype(common_dtype_),
+                                 LEGACY_CONTIGUOUS_MEMORY_FORMAT);
```

Comment:
Why not preserve current output format?

Reply:
Trying to preserve as much of the previous behavior as possible.